docs/contributor/adr/002-constructor-injection-and-composition-functions.md

```diff
@@ -0,0 +1,96 @@
+# ADR 002 - Constructor Injection and Composition Functions
+
+## Status
+
+Accepted
+
+## Context
+
+It is agreed that [Dependency Inversion](https://medium.com/@inzuael/solid-dependency-inversion-principle-part-5-f5bec43ab22e) is a key principle for solid architecture.
+Once a consumer abstracted their dependencies properly, the key question is how to fill the abstraction with a concrete implementation.
+
+It has to be decided how and where to resolve dependencies.
+
+## Decision
+
+It is decided that a consumer must have its dependencies injected via constructor.
+Constructors should follow the naming pattern `New<struct name>`.
+Stable dependencies of a consumer are resolved in a *composition function*.
+Composition functions live in a separate pacakge at `cmd/composition`.
+Composition functions should follow the naming pattern `Compose<struct name>`.
+Composition functions may build dependencies of the unit to compose themselves or have them injected from outside (depends on the kind of dependency).
```

Comment:
If I'm understanding correctly, the proposal is to create a composition function for every top-level dependency? That doesn't seem strictly necessary. For a service whose dependencies are all volatile, it can just be constructed directly in the root main without needing an extra composition layer, and I would say in most cases, use Construct is enough.